File: examples/catalog/money.go
package catalog

import (
	"fmt"
	"math/big"
)

type MoneySpec struct {
	Amount   string
	Currency string
}

type Money struct {
	amount   *big.Rat
	currency string
	_        [0]func()
}

func NewMoney(spec MoneySpec) (Money, error) {
	if spec.Currency == "" {
		return Money{}, fmt.Errorf("currency is required")
	}
	amount, ok := new(big.Rat).SetString(spec.Amount)
	if !ok {
		return Money{}, fmt.Errorf("invalid amount: %q", spec.Amount)
	}
	if amount.Sign() < 0 {
		return Money{}, fmt.Errorf("amount must not be negative: %q", spec.Amount)
	}
	return Money{amount: amount, currency: spec.Currency}, nil
}

func MustNewMoney(spec MoneySpec) Money {
	m, err := NewMoney(spec)
	if err != nil {
		panic(err)
	}
	return m
}

func (m Money) Currency() string { return m.currency }

func (m Money) Add(other Money) (Money, error) {
	if m.currency != other.currency {
		return Money{}, fmt.Errorf("cannot add %s and %s", m.currency, other.currency)
	}
	return Money{amount: new(big.Rat).Add(m.amount, other.amount), currency: m.currency}, nil
}

func (m Money) Equal(other Money) bool {
	return m.currency == other.currency && m.amount.Cmp(other.amount) == 0
}

func (m Money) String() string {
	return fmt.Sprintf("%s %s", m.amount.FloatString(2), m.currency)
}

File: examples/python/catalog/money.py
from dataclasses import dataclass
from decimal import Decimal, InvalidOperation


@dataclass(frozen=True)
class MoneySpec:

    amount: str
    currency: str


@dataclass(frozen=True, init=False)
class MoneyAmount:

    _value: Decimal

    def __init__(self, value: str) -> None:
        try:
            parsed = Decimal(value)
        except InvalidOperation as e:
            raise ValueError(f"invalid amount: {value!r}") from e
        if parsed < 0:
            raise ValueError(f"amount must not be negative: {parsed}")
        object.__setattr__(self, "_value", parsed)

    def add(self, other: "MoneyAmount") -> "MoneyAmount":
        return MoneyAmount(str(self._value + other._value))

    def __str__(self) -> str:
        return str(self._value)


@dataclass(frozen=True)
class MoneyCurrency:

    _value: str

    def __post_init__(self) -> None:
        if not self._value:
            raise ValueError("currency is required")

    def __str__(self) -> str:
        return self._value


@dataclass(frozen=True, init=False)
class Money:

    _amount: MoneyAmount
    _currency: MoneyCurrency

    def __init__(self, spec: MoneySpec) -> None:
        object.__setattr__(self, "_amount", MoneyAmount(spec.amount))
        object.__setattr__(self, "_currency", MoneyCurrency(spec.currency))

    @property
    def amount(self) -> MoneyAmount:
        return self._amount

    @property
    def currency(self) -> MoneyCurrency:
        return self._currency

    def add(self, other: "Money") -> "Money":
        if self._currency != other._currency:
            raise ValueError(f"cannot add {self._currency} and {other._currency}")
        total = self._amount.add(other._amount)
        return Money(MoneySpec(amount=str(total), currency=str(self._currency)))
